server/src/main/java/org/opensearch/gateway/ReplicaShardBatchAllocator.java

```diff
@@ -0,0 +1,175 @@
+     * match. Today, a better match is one that can perform a no-op recovery while the previous recovery
+     * has to copy segment files.
+     */
+    public void processExistingRecoveries(RoutingAllocation allocation, List<Set<ShardRouting>> shardBatches) {
+        RoutingNodes routingNodes = allocation.routingNodes();
+        List<Runnable> shardCancellationActions = new ArrayList<>();
+        for (Set<ShardRouting> shardBatch : shardBatches) {
+            Set<ShardRouting> eligibleFetchShards = new HashSet<>();
+            Set<ShardRouting> ineligibleShards = new HashSet<>();
+            for (ShardRouting shard : shardBatch) {
+                shardMatched = false;
+                for (RoutingNode routingNode : routingNodes) {
+                    if (routingNode.getByShardId(shard.shardId()) != null) {
+                        ShardRouting shardFromRoutingNode = routingNode.getByShardId(shard.shardId());
+                        if (!shardFromRoutingNode.primary()){
+                            shardMatched = true;
+                            if (shouldSkipFetchForRecovery(shardFromRoutingNode)) {
+                                continue;
+                            }
+                            eligibleFetchShards.add(shardFromRoutingNode);
+                        }
+                    }
+                    if (shardMatched){
+                        break;
+                    }
+                }
+            }
+            AsyncBatchShardFetch.FetchResult<NodeStoreFilesMetadataBatch> shardState = fetchData(eligibleFetchShards, ineligibleShards, allocation);
+            if (shardState.hasData() == false) {
+                logger.trace("{}: fetching new stores for initializing shard batch", eligibleFetchShards);
+                continue; // still fetching
+            }
+            for (ShardRouting shard : eligibleFetchShards) {
+                Map<DiscoveryNode, StoreFilesMetadata> nodeShardStores = getNodeShardStores(shard, shardState);
+
+                Runnable cancellationAction = getShardCancellationAction(shard, allocation, nodeShardStores);
+                if (cancellationAction != null) {
+                    shardCancellationActions.add(cancellationAction);
+                }
+            }
+        }
+        for (Runnable action : shardCancellationActions) {
+            action.run();
+        }
+    }
+
+
+    abstract protected FetchResult<NodeStoreFilesMetadataBatch> fetchData(Set<ShardRouting> shardEligibleForFetch,
+                                                                          Set<ShardRouting> inEligibleShards,
+                                                                          RoutingAllocation allocation);
+
+    @Override
+    protected FetchResult<TransportNodesListShardStoreMetadata.NodeStoreFilesMetadata> fetchData(ShardRouting shard, RoutingAllocation allocation) {
+        return null;
```

Comment:
Let us add a Log Message here. 

Would `IlleagaStateException` be a better action here than returning `null` ?

Reply:
+1

Reply:
Updated in latest commit